# Data Consistency Checks

# 01. Importing Libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Original Data with a path creation

In [2]:
path = r'/Users/ladyarmanoid/Documents/Data Analytics/CareerFoundry/Course/2- Data Immersion/Achievement 4/Instacart Basket Analysis/'

# Products Dataframe Import

In [3]:
# df_prods = Products data frame
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [3]:
# df_prods = Products data frame
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

# Orders Dataframe Import

In [4]:
# df_ords = Orders data frame
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

In [5]:
# Create a data frame
df_test = pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# 03. Check for Mixed Data Types

In [8]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


Fixing the mix-type columns offers two options generally: going 'str' or 'int64'
The code for a string here is for example df_test['mix'] = df_test['mix'].astype('str')
The code for a numerical value is df_test['mix'] = df_test['mix'].astype('int64')

In [9]:
# Fixing the mix-type df_test data frame to become a string
df_test['mix'] = df_test['mix'].astype('str')

# 04. Finding Missing Values

In [10]:
# Check if df_prods comes with missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

## Create a Subset to view the 16 missing values

In [11]:
# Create df_nan, only containing the product_name column that meet the condition isnull() = True
df_nan = df_prods [df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## What logic to use when imputing missing data -> numeric values only

Use df.describe () and use the following code to replace your missing values with the mean value
df['column with missings'].fillna(mean value, inplace=True)

If you chose to use the median, you reuse the same above function with the median value instead
Use the df.median() then use df['column with missings'].fillna(median value, inplace=True)

Another alternative is using linear interpolation by connecting two points with a line

# 05. Imputation for String Values

In [13]:
df_prods.shape # you compare here the number of rows in your current df_prods with the number in your subset df_nan

(49693, 5)

In [14]:
# New df to harmonize the data frame
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

Another way to drop all missing values: df_prods.dropna(inplace = True)
If you only one to drop the NaNs: df_prods.dropna(subset = [‘product_name’], inplace = True)

# 06. Finding Duplicates

In [16]:
# Function to look for full duplicates within df_prods
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 07. Addressing Duplicates

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
# New df excluding duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task Steps - Products and Orders Dataframes

## Check Data Consistency

In [21]:
# Step 1. Run the df.describe() function on your df_prods dataframe:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The minimum value for 'prices' is 1 and the maximum value is 99,999, which seems to be excessive. There could be an error

In [22]:
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

The price listed for the '2% Reduced Fat Milk' need to be corrected. Contacting the client for an updated data frame is necessary 

In [23]:
# Step 2. Run the df.describe() function on your df_ords dataframe:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

1- the days_since_prior_order has a minimum value of 0 when the maximum is 3 days. But we can see that the percentile range is not logical as we have a 50th percentile of 7, higher than the max. value

2 - the number of orders also has a min of 1 and a max of 1, although the percentile ranges are inconsistent with this. For instance, 25% is recorded as 5, higher than the max of 1.

In [24]:
df_ords.head() # Check for mixed data types

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [25]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

*Step 3:* **no mix data were found in df_ords**

# Checking for Missing Values

In [26]:
# Check if df_ords comes with missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

*Step 5:* **we find 206,209 missing values for the column "days_since_prior_order"**

## Create a Subset to view the 206209 missing values

In [27]:
# Create df_ords_nan, only containing the days_since_prior_order column that meet the condition isnull() = True
df_ords_nan = df_ords [df_ords['days_since_prior_order'].isnull() == True]

In [28]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

## The missing values only mean that the order registered was the first order done by the unique 'user_id' customer.
**For this reason, there will be no changes done to the data**

In [29]:
df_ords.shape # you compare here the number of rows in your current df_prods with the number in your subset df_ords_nan just for info

(3421083, 8)

# Finding and Addressing Duplicates

In [31]:
# Function to look for full duplicates within df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [32]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

*Step 7:* **no duplicates found in df_ords**
*Step 8:* **there are no actions needed as no duplicates were found in df_ords**

In [33]:
# Exporting the products and  orders data frames
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))
